common/core/overlaylayer.cpp

```diff
@@ -229,55 +229,45 @@ void OverlayLayer::InitializeState(HwcLayer* layer,
-    if (display_frame_.right > right_constraint) {
-      display_frame_.right = right_constraint;
-    }
+  int32_t left_source_constraint = layer->GetLeftSourceConstraint();
+  int32_t right_source_constraint = layer->GetRightSourceConstraint();
+  int32_t display_frame_left = display_frame_.left;
+  uint32_t frame_width = display_frame_.right - display_frame_.left;
+  uint32_t source_width = source_crop_.right - source_crop_.left;
+  uint32_t frame_offset_left = 0;
+  uint32_t frame_offset_right = frame_width;
 
-    if (display_frame_.left < left_constraint) {
-      display_frame_.left = left_constraint;
-    }
-
-    if (display_frame_.right < right_constraint) {
-      display_frame_.right =
-          std::max(display_frame_.left, display_frame_.right);
-    }
+  if (left_constraint >= 0 && right_constraint >= 0) {
+    if (display_frame_.right < left_source_constraint ||
+        display_frame_.left > right_source_constraint) {
+      display_frame_.left = 0;
+      display_frame_.right = 0;
+      frame_offset_right = 0;
+    } else {
+      if (display_frame_.left < left_source_constraint) {
+        frame_offset_left = left_source_constraint - display_frame_left;
+        display_frame_.left = left_source_constraint;
+      }
 
-    if (display_frame_.left > left_constraint) {
-      display_frame_.left = std::min(display_frame_.left, display_frame_.right);
+      if (display_frame_.right > right_source_constraint) {
+        frame_offset_right = right_source_constraint - display_frame_left;
+        display_frame_.right = right_source_constraint;
+      }
     }
 
-    if (left_constraint > 0) {
-      display_frame_.left -= left_constraint;
-      display_frame_.right = right_constraint - display_frame_.right;
-    }
+    display_frame_.left =
+        display_frame_.left - left_source_constraint + left_constraint;
+    display_frame_.right =
+        display_frame_.right - left_source_constraint + left_constraint;
 
     display_frame_.bottom =
         std::min(max_height, static_cast<uint32_t>(display_frame_.bottom));
     display_frame_width_ = display_frame_.right - display_frame_.left;
     display_frame_height_ = display_frame_.bottom - display_frame_.top;
     surface_damage_ = display_frame_;
-  }
-
-  float lconstraint = (float)layer->GetLeftSourceConstraint();
-  float rconstraint = (float)layer->GetRightSourceConstraint();
-  if (lconstraint >= 0 && rconstraint >= 0) {
-    if (source_crop_.right > rconstraint) {
-      source_crop_.right = rconstraint;
-    }
-
-    if (source_crop_.left < lconstraint) {
-      source_crop_.left = lconstraint;
-    }
-
-    if (source_crop_.right < rconstraint) {
-      source_crop_.right = std::max(source_crop_.left, source_crop_.right);
-    }
-
-    if (source_crop_.left > lconstraint) {
-      source_crop_.left = std::min(source_crop_.left, source_crop_.right);
-    }
-
+    source_crop_.left = source_width * ((float)frame_offset_left / frame_width);
```

Comment:
hmm not sure why we need to do this. can you explain. We shouldnt manipulate source crop rect in any way.  I think right way to fix this would be solving it in mosaic to send left and right constraints similar to source left and right constraints ? Did u try that. Any thoughts

Reply:
as the original source crop rect included the contend which crossed more than one sub displays of mosaic display, we need split it for a special sub display.

Reply:
k, can u explain what the formula does here, or better add it as comments in source code.